java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImpl.java

```diff
@@ -139,31 +143,87 @@ public String getRegionSuffix() {
     return regionSuffix;
   }
 
-  public static final class Builder implements BuildStep {
+  String getTableName() {
+    return tableName;
+  }
+
+  DynamoDB getClient() {
+    return client;
+  }
+
+  public static final class Builder implements BuildStep, EndPointStep, RegionStep {
 
-    private String regionSuffix = "";
     private AmazonDynamoDB client;
 
+    private String regionSuffix = "";
+    private AmazonDynamoDBClientBuilder standardBuilder = AmazonDynamoDBClientBuilder.standard();
+    private String tableName = "EncryptionKey";
+
+
+    @Override
+    public BuildStep withEndPointConfiguration(final String endPoint, final String signingRegion) {
+      standardBuilder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endPoint, signingRegion));
+      return this;
+    }
+
+    @Override
+    public BuildStep withTableName(final String table) {
+      this.tableName = table;
+      return this;
+    }
+
     @Override
-    public BuildStep withDynamoDbRegionSuffix(final String region) {
-      this.regionSuffix = region;
+    public BuildStep withDynamoDbRegionSuffix(final String suffix) {
+      this.regionSuffix = suffix;
+      return this;
+    }
+
+    @Override
+    public BuildStep withRegion(final String region) {
+      standardBuilder.withRegion(region);
       return this;
     }
 
     @Override
     public DynamoDbMetastoreImpl build() {
-      client = AmazonDynamoDBClientBuilder.standard().build();
-      return new DynamoDbMetastoreImpl(new DynamoDB(client), regionSuffix);
+      client = standardBuilder.build();
+      return new DynamoDbMetastoreImpl(this);
     }
   }
 
+  public interface EndPointStep {
+    /**
+     * Adds EndPoint config to the AWS DynamoDb client
+     * @param endPoint the service endpoint either with or without the protocol
+     * @param signingRegion the region to use for SigV4 signing of requests (e.g. us-west-1)
+     * @return The current {@code BuildStep} instance
+     */
+    BuildStep withEndPointConfiguration(String endPoint, String signingRegion);
+  }
+
+  public interface RegionStep {
+    /**
+     * Specifies the region for the AWS DynamoDb client
+     * @param region
+     * @return The current {@code BuildStep} instance
+     */
+    BuildStep withRegion(String region);
+  }
+
   public interface BuildStep {
     /**
      * Specifies whether region suffix should be enabled for DynamoDB
-     * @param region The region to be used as suffix
+     * @param suffix The region to be used as suffix
+     * @return The current {@code BuildStep} instance.
+     */
+    BuildStep withDynamoDbRegionSuffix(String suffix);
+
+    /**
+     * Specifies the name of the table
+     * @param table The name of the table
      * @return The current {@code BuildStep} instance.
      */
-    BuildStep withDynamoDbRegionSuffix(String region);
+    BuildStep withTableName(String table);
```

Comment:
this step should be modified to use kind of a step builder.
currently, i can create a `DynamoDbMetastoreImpl` without specifying a table name, hence causing an error.

Having a step builder would force you to specify a table name before calling the actual `build()` step.

Reply:
Creating a `DynamoDbMetastoreImpl` object without a table name does not cause an error.
The table name defaults with `EncryptionKey` if the `withTableName` builder step isn't  called. https://github.com/sushantmimani/asherah/blob/866d6ccc20019257494bed33b04e158484be8f98/java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImpl.java#L160

All our existing tests run without this step
https://github.com/sushantmimani/asherah/blob/866d6ccc20019257494bed33b04e158484be8f98/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImplTest.java#L73

Reply:
Updated the code to be more specific regarding default values